CNV- 10120: dynamic ssh keys (#63313)
openshift/openshift-docs: modules/virt-about-static-and-dynamic-ssh-keys.adoc

== Dynamic SSH key management

You can enable dynamic SSH key management for a VM with {op-system-base-full} 9 installed. Afterwards, you can update the key during runtime. The key is added by the QEMU guest agent, which is installed with Red Hat boot sources.

You can disable dynamic key management for security reasons. Then, the VM inherits the key management setting of the image from which it was created.

.Use cases

* Granting or revoking access to VMs: As a cluster administrator, you can grant or revoke remote VM access by adding or removing the keys of individual users from a `Secret` object that is applied to all VMs in a namespace.

* User access: You can add your access credentials to all VMs that you create and manage.

* Ansible provisioning:

** As an operations team member, you can create a single secret that contains all the keys used for Ansible provisioning.
** As a VM owner, you can create a VM and attach the keys used for Ansible provisioning.

* Key rotation:

** As a cluster administrator, you can rotate the Ansible provisioner keys used by VMs in a namespace.
** As a workload owner, you can rotate the key for the VMs that you manage.

> 

Reviewer: ```suggestion
```

// Module included in the following assemblies:
//
// * virt/virtual_machines/virt-accessing-vm-ssh.adoc

:_content-type: REFERENCE
[id="virt-about-static-and-dynamic-ssh-keys_{context}"]
= About static and dynamic public keys

You can add static or dynamic public SSH keys to virtual machines (VMs).

> 

Reviewer 2: Static and dynamic are not attributes of the public key, but attributes of the way the key is added the VM.

```suggestion
You can add public SSH keys to virtual machines (VMs) statically at the first boot or dynamically at runtime.
```

[discrete]
[id="static-public-key_{context}"]
== Static public key

> A static public SSH key is added to the VM at startup. It is propagated by using a cloud-init configuration drive and does not affect cloud-init user data.

Reviewer 2: I am not sure of `start up` or `first boot` fits better.
```suggestion
A static public SSH key is added to the VM at startup. It is propagated by using a cloud-init data source.
```

> All operating systems support static public keys.

Reviewer 2: It is required that the guest is able to apply the configuration from the cloud-init data soruce. This is the case for the RHEL cloud image, but also many 3rd party cloud images. It is even possible to use cloud-init in windows, e.g. via https://cloudbase.it/cloudbase-init/ .

You can add a static public key by using one of the following methods:

* Add a key to the VM when you create it by using the web console or the command line.
* Add a key to a project by using the web console. Afterwards, the key is automatically added to the VMs that you create in this project.

.Use cases

> * As a VM owner, you can provision all your VMs in a project with static keys.

Reviewer 2: ```suggestion
* As a VM owner, you like to provision all your newly created VMs to accept a public ssh key.
```

[discrete]
[id="dynamic-public-key_{context}"]
== Dynamic public key

> A dynamic public SSH key can be added or revoked at any time. It is propagated by the QEMU guest agent, which is included in Red Hat templates.

Reviewer 2: ```suggestion
A dynamic public SSH key can be added or revoked while the VM is running. It is propagated by the QEMU guest agent, which is included in Red Hat default boot sources.
```

> Only {op-system-base-full} 9 supports dynamic public keys.

Reviewer 2: ```suggestion
Only {op-system-base-full} 9 supports managing the public keys dynamically.
```
